File: SuperLauncherCpp/src/ui/pages/ServersPage.cpp
#include "ServersPage.h"
#include "src/backend/ServerManager.h"
#include <QHBoxLayout>

ServersPage::ServersPage(ServerManager *mgr, QWidget *parent)
    : QWidget(parent), m_mgr(mgr)
{
    setupUi();
}

void ServersPage::setupUi()
{
    auto *mainLayout = new QVBoxLayout(this);
    mainLayout->setContentsMargins(24, 24, 24, 24);

    auto *header = new QHBoxLayout();
    auto *title = new QLabel("Серверы", this);
    title->setStyleSheet("color: white; font-size: 22px; font-weight: bold;");

    auto *createBtn = new QPushButton("+ Создать сервер", this);
    createBtn->setStyleSheet(
        "QPushButton { background: #4CAF50; color: white; border: none; "
        "border-radius: 6px; padding: 8px 20px; font-size: 13px; }"
        "QPushButton:hover { background: #45a049; }");
    connect(createBtn, &QPushButton::clicked, this, [this]() {
        // Will open dialog
    });

    header->addWidget(title);
    header->addStretch();
    header->addWidget(createBtn);
    mainLayout->addLayout(header);

    // Scrollable cards
    m_scrollArea = new QScrollArea(this);
    m_scrollArea->setWidgetResizable(true);
    m_scrollArea->setStyleSheet("QScrollArea { border: none; background: transparent; }");

    auto *scrollContent = new QWidget(m_scrollArea);
    m_cardsLayout = new QVBoxLayout(scrollContent);
    m_cardsLayout->setSpacing(12);
    m_cardsLayout->addStretch();
    m_scrollArea->setWidget(scrollContent);
    mainLayout->addWidget(m_scrollArea, 1);
}

void ServersPage::refreshServers(const QString &serversBase)
{
    m_serversBase = serversBase;
    // Clear existing cards (except stretch)
    QLayoutItem *item;
    while ((item = m_cardsLayout->takeAt(0))) {
        if (item->widget())
            item->widget()->deleteLater();
        delete item;
    }

    auto servers = m_mgr->getServers(serversBase);
    for (const auto &s : servers)
        addServerCard(s);
    m_cardsLayout->addStretch();
}

void ServersPage::addServerCard(const Server &server)
{
    auto *card = new ServerCardWidget(server, this);

    connect(card, &ServerCardWidget::startClicked, this,
            &ServersPage::startServer);
    connect(card, &ServerCardWidget::stopClicked, this,
            &ServersPage::stopServer);
    connect(card, &ServerCardWidget::removeClicked, this,
            [this](const Server &s) { emit removeServer(s.name); });

    m_cardsLayout->insertWidget(m_cardsLayout->count() - 1, card);
}

File: SuperLauncher/ui/pages/servers_page.py
import os
from PyQt6.QtWidgets import (QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton,
                              QLineEdit, QScrollArea, QFrame, QMessageBox)
from PyQt6.QtCore import Qt
from PyQt6.QtGui import QCursor

class ServersPage(QWidget):
    def __init__(self, main):
        super().__init__()
        self.main = main
        layout = QVBoxLayout(self)
        layout.setContentsMargins(15, 15, 15, 15)

        title = QLabel("Серверы")
        title.setStyleSheet("font-size: 26px; font-weight: bold; color: white;")
        layout.addWidget(title)

        create_btn = QPushButton("Создать сервер")
        create_btn.clicked.connect(self._create)
        layout.addWidget(create_btn)

        row = QHBoxLayout()
        self.name_input = QLineEdit()
        self.name_input.setPlaceholderText("Имя сервера")
        row.addWidget(self.name_input)
        self.ip_input = QLineEdit()
        self.ip_input.setPlaceholderText("IP или домен")
        row.addWidget(self.ip_input)
        add_btn = QPushButton("Добавить")
        add_btn.clicked.connect(self._add)
        row.addWidget(add_btn)
        layout.addLayout(row)

        scroll = QScrollArea()
        scroll.setWidgetResizable(True)
        self.container = QWidget()
        self.container_layout = QVBoxLayout(self.container)
        scroll.setWidget(self.container)
        layout.addWidget(scroll, 1)

        self._refresh()

    def _refresh(self):
        for i in reversed(range(self.container_layout.count())):
            w = self.container_layout.itemAt(i).widget()
            if w: w.deleteLater()

        servers = self.main.servers.load_servers()
        for s in servers:
            card = QFrame()
            card.setStyleSheet("QFrame { background: rgba(40,40,55,0.9); border: 1px solid rgba(255,255,255,0.1); border-radius: 10px; padding: 10px; margin: 2px; }")
            cl = QHBoxLayout(card)
            cl.addWidget(QLabel(f"<b>{s['name']}</b>", styleSheet="font-size: 15px; color: white;"))
            cl.addWidget(QLabel(s.get('ip',''), styleSheet="color: #4facfe;"))
            cl.addStretch()
            if s.get('managed'):
                console_btn = QPushButton("Консоль")
                console_btn.clicked.connect(lambda checked, n=s['name']: self._console(n))
                cl.addWidget(console_btn)
            del_btn = QPushButton("Удалить")
            del_btn.clicked.connect(lambda checked, n=s['name']: self._delete(n))
            cl.addWidget(del_btn)
            self.container_layout.addWidget(card)
        self.container_layout.addStretch()

    def _add(self):
        name = self.name_input.text().strip()
        ip = self.ip_input.text().strip()
        if not name or not ip:
            QMessageBox.warning(self, "Ошибка", "Заполните поля")
            return
        servers = self.main.servers.load_servers()
        servers.append({"name": name, "ip": ip, "managed": False})
        self.main.servers.save_servers(servers)
        self._refresh()
        self.name_input.clear()
        self.ip_input.clear()

    def _create(self):
        from ..dialogs.create_server_dialog import CreateServerDialog
        dlg = CreateServerDialog(self.main)
        if dlg.exec():
            self._refresh()

    def _console(self, name):
        from ..dialogs.server_control_dialog import ServerControlDialog
        path = os.path.join("servers", name)
        dlg = ServerControlDialog(name, path, self.main)
        dlg.exec()

    def _delete(self, name):
        reply = QMessageBox.question(self, "Удаление", f"Удалить сервер '{name}'?",
                                      QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No)
        if reply == QMessageBox.StandardButton.Yes:
            servers = self.main.servers.load_servers()
            servers = [s for s in servers if s['name'] != name]
            self.main.servers.save_servers(servers)
            path = os.path.join("servers", name)
            if os.path.exists(path):
                import shutil
                shutil.rmtree(path, ignore_errors=True)
            self._refresh()
